store/s3store.go

```diff
@@ -0,0 +1,170 @@
+package store
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+const Suffix string = ".json"
+
+type S3Store struct {
+	Bucket string
+	Prefix string
+	svc    s3.S3
+}
+
+// ensure S3Store conforms to Store interface
+var _ Store = &S3Store{}
+
+// Establish an AWS Session and deduce the region:
+// - from $CHAMBER_AWS_REGION
+// - from EC2 metadata
+func awsSessionAndRegion() (*session.Session, *string) {
+	var region *string
+
+	if regionOverride, ok := os.LookupEnv("CHAMBER_AWS_REGION"); ok {
+		region = aws.String(regionOverride)
+	}
+	sess := session.Must(session.NewSessionWithOptions(
+		session.Options{
+			Config: aws.Config{
+				Region: region,
+			},
+			SharedConfigState: session.SharedConfigEnable,
+		},
+	))
+
+	// If region is still not set, attempt to determine it via ec2 metadata API
+	region = nil
+	if aws.StringValue(sess.Config.Region) == "" {
+		sessEc2 := session.New()
+		ec2metadataSvc := ec2metadata.New(sessEc2)
+		if regionOverride, err := ec2metadataSvc.Region(); err == nil {
+			region = aws.String(regionOverride)
+		}
+	}
+	return sess, region
+}
+
+func NewS3Store(numRetries int, bucket string, prefix string) *S3Store {
+	session, region := awsSessionAndRegion()
+
+	svc := s3.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+	s := S3Store{
+		svc:    *svc,
+		Bucket: bucket,
+		Prefix: prefix,
+	}
+
+	return &s
+}
+
+func (s *S3Store) surround(id string) string {
+	return s.Prefix + id + Suffix
+}
+
+// Return's Secrets.Meta.LastModifed will be time.Time 0-value
+func (s *S3Store) WriteAll(id string, secrets RawSecrets) (string, error) {
```

Comment:
we'll want to figure out how we want to handle encryption. use chamber for chamber to store the secret private key?